### PR TITLE
Added bootstrap_.py that takes filename as argument along with LOA and year

### PR DESCRIPTION
**Pull request by noorains**

### CODE REVIEW

**Comment by nskazmi** [comment]
image

**Comment by nskazmi** [comment]
this is how it works from terminal

File: bootstrap_.py
Comment (kvelleby) [inline bootstrap_.py:1]: I was thinking you would be better off using the features data than the "actuals" files. Here is a short example of what I was thinking. It should then be saved to a Apache Hive kind of structure: i.e., benchmark_boostrap/pgm/year=2018/predictions.parquet ... and so on. The implementation now uses "the year before". One question is whether we should use the year before, but from Oct-Oct. E.g., for 2020, we draw from a pool of data from Oct 2018 to Oct 2019. The features data now contains a "month" column so you could select that (pac.field("year) > year-2, then subset the pandas file removing the months that should not be in the pool.

```
def global_bootstrap_benchmark(feature_folder, target, year) -> pd.DataFrame:
    """
    Draws a random sample with replacement from the complete global pool of fatalities data from the year before.
    """
    
    if target == "pgm":
        unit = "priogrid_gid"
    elif target == "cm":
        unit = "country_id"
    else:
        raise ValueError('Target must be "pgm" or "cm.')


    filter = pac.field("year") == year - 1
    pool = pq.ParquetDataset(feature_folder / target, filters = filter).read(columns = ["ged_sb"]).to_pandas()

    filter = pac.field("year") == year
    df = pq.ParquetDataset(feature_folder / target, filters = filter).read(columns = [unit, "month_id"]).to_pandas()

    df["outcome"] = np.random.choice(pool["ged_sb"], size = (df.shape[0], 1000), replace = True).tolist()
    df = df.explode('outcome').astype('int32')
    df['draw'] = df.groupby(['month_id', unit]).cumcount()
    return df



feature_folder = Path("path/to/features")
targets = ["pgm", "cm"]
years = range(2018, 2022)
for target in targets:
    for year in years:
        global_bootstrap_benchmark(feature_folder, target, year)
```

Comment (kvelleby) [reply bootstrap_.py:1]: Some of the things you have written here (like parsing month_id's) etc, I have put in the utilities.py module

Comment (kvelleby) [reply bootstrap_.py:1]: Great that you add a command-line interface. If you rename the module to "benchmarks.py", then the main() should have an argument for the kind of benchmark to run, and the path to the feature_folder. The rest of the arguments should have sensible defaults (like selecting both cm and pgm level, and iterating over relevant years).

**Review by kvelleby — CHANGES_REQUESTED** [review]
Thanks for this. Your code works, but I would like some refactoring, partially due to a moving target from my side. Sorry about that. See comments in the boostrap_.py module. Other than those comments, I also want a request containing only this file, without the ipython notebook. Thanks! :)

File: bootstrap_.py
Comment (kvelleby) [reply bootstrap_.py:1]: Also note: 
```
import pyarrow.compute as pac
import pyarrow.parquet as pq
```

**Pull request closed without merging**